### PR TITLE
WIP: Added new LupIO-TTY device 

### PR DESCRIPTION
**Pull request by lhinman19**

### CODE REVIEW

File: src/dev/lupio/lupio_tty.cc
Comment (powerjg) [inline src/dev/lupio/lupio_tty.cc:54]: I wouldn't have multiple assignments on a single line. Also, for class member variables the initialization should be in the initializer list. OR, you can actually initialize in the header if the initial value is a constant.

File: src/dev/lupio/lupio_tty.cc
Comment (powerjg) [inline src/dev/lupio/lupio_tty.cc:70]: Maybe just have this call once at the end of the function?

File: src/dev/lupio/lupio_tty.cc
Comment (powerjg) [inline src/dev/lupio/lupio_tty.cc:78]: What is `r`? This variable name is too short (and it's unclear if it's a local variable or class variable).

**Review by powerjg — COMMENTED** [review]
A few quick comments :)

File: src/dev/riscv/LupV.py
Comment (lhinman19) [inline src/dev/riscv/LupV.py:187]: I still have been unable to get this to work with "earlycon console=ttyLIO0". As Jason mentioned, if I use kernel parameters "earlycon=lupio_tty,0x20007000" then the earlyconsole does work.

File: src/dev/lupio/LupioTTY.py
Comment (joel-porquet) [inline src/dev/lupio/LupioTTY.py:49]: Back to our convo from the other day, it does seem like this unproxying approach offers a certain level of genericity.

If the platform defined its interrupt controller with a more generic name (eg `intc` or `pic`), then this component (and same for the RNG) could have a line such as `pic = platform.pic`.

The only constraint for using the lupio devices from any platform would then to be that the platform defines a member called `pic`, but the interrupt controller itself could be of any kind (instead of implicitly restricting it to be a PLIC necessarily like the naming here strongly suggests).

In other words, until the new FDT code lands, I suggest to use this unproxying approach but to replace `plic` with `pic`.

File: src/dev/lupio/lupio_tty.cc
Comment (joel-porquet) [inline src/dev/lupio/lupio_tty.cc:61]: You should actually read the data here: `readChar = device->readData();`. That's what makes everything else work (e.g., returning a correct value for the STAT register)

File: src/dev/lupio/lupio_tty.cc
Comment (joel-porquet) [inline src/dev/lupio/lupio_tty.cc:63]: Call `lupioTTYUpdateIRQ()` from here, that's what this function is for.

File: src/dev/lupio/lupio_tty.cc
Comment (joel-porquet) [inline src/dev/lupio/lupio_tty.cc:81]: The clearing of the INT should be uncommented.

File: src/dev/lupio/lupio_tty.cc
Comment (joel-porquet) [inline src/dev/lupio/lupio_tty.cc:100]: Drop that entire block.

File: src/dev/lupio/lupio_tty.cc
Comment (joel-porquet) [inline src/dev/lupio/lupio_tty.cc:109]: Remove this line, the call to the update function should take care of it.

File: src/dev/lupio/lupio_tty.cc
Comment (joel-porquet) [inline src/dev/lupio/lupio_tty.cc:127]: Indentation.

**Review by joel-porquet — CHANGES_REQUESTED** [review]
Try with what I suggested, I believe the input should work after that. :crossed_fingers: 

Generally, it seems like the more you stick to the same approach as the QEMU implementation, the higher the chances it will work.

File: src/dev/riscv/LupV.py
Comment (joel-porquet) [inline src/dev/riscv/LupV.py:188]: It should technically be "/soc/lupio-tty@20007000" (for some reason, addresses in node names do not contain the `0x` prefix)

File: src/dev/riscv/lupv.cc
Comment (joel-porquet) [inline src/dev/riscv/lupv.cc:57]: You can't raise interrupts for both interrupts. Otherwise, the software will be completely confused. If the uart calls `postConsoleInt()` then only the related IRQ at the PLIC level should be raised, and same for the lupio-tty.

You can either define another set of postConsoleInt()/clearConsoleInt() (eg with a different naming) that is just for the lupio-tty, or you need to accept an argument that says specifically which one should be raised.

I would suggest making a pair of generic functions, such as `postPicInt/clearPicInt(int intID)`. If each device knows its own intID, then it can call this pair of functions with it. It will be useful for lupio-blk as well.

File: src/dev/lupio/LupioTTY.py
Comment (powerjg) [reply src/dev/lupio/LupioTTY.py:49]: That's fine with me. Both approaches are awful in their own way. :)

File: src/dev/lupio/lupio_tty.cc
Comment (powerjg) [reply src/dev/lupio/lupio_tty.cc:127]: And curly braces... I wish I could change our style to *require* them. I'm happy to show you examples of how skipping curly braces caused future bugs :)

File: src/dev/riscv/lupv.cc
Comment (powerjg) [reply src/dev/riscv/lupv.cc:57]: Something I don't understand about the platform code in general: Is there a reason the lupioTTY has to call into the platform to post the interrupt to the plic? Could it just have a pointer to a generic device and call the post() function itself?

Comment (joel-porquet) [reply src/dev/riscv/lupv.cc:57]: I guess one drawback of having the device call the `post` function of the interrupt controller directly is that the device needs a pointer to the interrupt controller instance with a type that defines the `post` function. In that case, we're back to the problem of the device being dependent on a certain model of interrupt controller.

A possible solution that may solve this is to have a proper hierarchy of classes. If interrupt controllers derived from a generic interrupt controller class that defines a pair of `post/clear` virtual functions, then devices would only need to have a pointer onto an instance of this generic class and call these functions without any issues.

Or does the BasicPioDevice already define a `post()` function?

Comment (powerjg) [reply src/dev/riscv/lupv.cc:57]: Ah, that makes sense. No, it doesn't look like there's a virtual `post()` function, which is unfortunate.

I guess you need to add another `LupV::post???Interrupt` function for the lupio TTY. Maybe use `postUartInt()` instead of "console"?

File: src/dev/lupio/lupio_tty.cc
Comment (powerjg) [inline src/dev/lupio/lupio_tty.cc:79]: I don't understand how this can possibly be "generic" between ISAs right now. The platform is ISA (or really *board*) specific. The generic platform (`src/dev/platform.hh`) only has a `postConsoleInterrupt` and a `postPciInt`. I don't know how you're supposed to use these functions.

File: src/dev/riscv/lupv.cc
Comment (lhinman19) [reply src/dev/riscv/lupv.cc:57]: Sorry, I left the uart in there because I was switching out the uart and LupIO-TTY when testing and I had intended to remove the uart entirely from the LupV platform. Should I then support both the uart8250 and LupIO-TTY in the LupV platform?

Comment (powerjg) [reply src/dev/riscv/lupv.cc:57]: Yeah, that's fine for the LupV platform, but now I'm curious as to how this is supposed to work more generally! What if there's two consoles? (I don't think that's impossible in real life...)

File: src/dev/lupio/lupio_tty.cc
Comment (lhinman19) [reply src/dev/lupio/lupio_tty.cc:81]: I'm a little concerned that when I clear the INT here, in the Linux boot-up phase I'm getting many "clear interrupt" statements, seemingly without having posted any interrupts. However, everything does seem to work properly and when I was looking at the uart8250 it also seems to output several `cleared` statements without the corresponding interrupt statements.

File: src/dev/lupio/LupioTTY.py
Comment (lhinman19) [reply src/dev/lupio/LupioTTY.py:48]: Should I still make this change now that we have the new FDT code?

Comment (powerjg) [reply src/dev/lupio/LupioTTY.py:48]: Yeah, the new code is going to take a few days to get in, and then you'll have to rebase. I vote that you shouldn't be blocked by that :)

File: src/dev/riscv/lupv.hh
Comment (lhinman19) [inline src/dev/riscv/lupv.hh:50]: I'm not sure if this is what was requested. The pic is still a type `Plic` because I wasn't sure if I was supposed to create a generic class named `Pic` from which the `plic` can be derived?

File: src/dev/lupio/LupioTTY.py
Comment (joel-porquet) [inline src/dev/lupio/LupioTTY.py:38]: We may want to find a better name than `device` for this parameter (and then, for the class member). What about `serial` or `terminal`?

File: src/dev/lupio/lupio_tty.cc
Comment (joel-porquet) [inline src/dev/lupio/lupio_tty.cc:75]: I wonder if that should really be an `if`. If this function is called, it *should* be because there is data available. The lack of data would then be an error, caught as such:

```C
assert(device->dataAvailable());
```

File: src/dev/lupio/lupio_tty.cc
Comment (joel-porquet) [inline src/dev/lupio/lupio_tty.cc:68]: The reason for the "clear interrupt" to be printed when the interrupt is actually not *unset* (ie go from high to low) is that this function may be called to operate an update that ends up not actually changing the state of the IRQ. It's more the printf that doesn't report the correct event. You can either remove the printf, or change what it says in order to remove any confusion.

File: src/dev/riscv/LupV.py
Comment (joel-porquet) [inline src/dev/riscv/LupV.py:106]: Why not keep the same interrupt ID as the uart? (it's a question)

File: src/dev/riscv/LupV.py
Comment (joel-porquet) [inline src/dev/riscv/LupV.py:110]: Extra space?

File: src/dev/riscv/lupv.cc
Comment (joel-porquet) [inline src/dev/riscv/lupv.cc:67]: In a separate PR, we could probably remove everything related to PCI.

File: src/dev/riscv/lupv.hh
Comment (joel-porquet) [reply src/dev/riscv/lupv.hh:50]: It'll get replaced by the lupio-pic eventually, so it's not a big deal.

**Review by joel-porquet — CHANGES_REQUESTED** [review]
Great progress! Does it all work in the end?

File: src/dev/lupio/lupio_tty.cc
Comment (powerjg) [reply src/dev/lupio/lupio_tty.cc:75]: `chatty_assert` is preferred: https://gem5.googlesource.com/public/gem5/+/refs/heads/develop/src/base/logging.hh#304

File: src/dev/riscv/LupV.py
Comment (lhinman19) [reply src/dev/riscv/LupV.py]: No good reason, except when I was examining the device tree in QEMU I think the interrupt id for lupio-tty was 0xb but I just changed it back to 0xa

**Comment by lhinman19** [comment]
Yes, I was having problems running this code with `telnet` but once I switched over to the `m5term` everything has been working just fine :)

File: src/dev/lupio/lupio_tty.cc
Comment (powerjg) [reply src/dev/lupio/lupio_tty.cc]: Actually, now it's just `gem5_assert` if you rebase on current develop: https://gem5-review.googlesource.com/c/public/gem5/+/48606

File: src/dev/lupio/lupio_tty.cc
Comment (lhinman19) [reply src/dev/lupio/lupio_tty.cc]: I think r stands for return value? I can definitely change this naming but we may also need to change it in the RNG and RTC since I think they also use "r".

File: src/dev/lupio/lupio_tty.cc
Comment (joel-porquet) [reply src/dev/lupio/lupio_tty.cc:65]: The generic platform does seem to have a virtual function called `postConsoleInt()` (https://github.com/darchr/gem5/blob/develop/src/dev/platform.hh#L62). So I guess calling `platform->postConsoleInt()` should work regardless of the platform, no?

File: src/dev/lupio/lupio_tty.cc
Comment (joel-porquet) [reply src/dev/lupio/lupio_tty.cc:119]: Do you require braces for single statements as well?

File: src/dev/lupio/lupio_tty.cc
Comment (joel-porquet) [reply src/dev/lupio/lupio_tty.cc]: FWIW, `r` is the variable name typically used in QEMU for returning a value in devices' read function. It probably stands for "return value" or "register". That's the name I also used in my QEMU models. If it doesn't fit gem5's coding style, we could use something like `ret`.

**Review by joel-porquet — COMMENTED** [review]
Looks good to me!

File: src/dev/lupio/lupio_tty.cc
Comment (powerjg) [reply src/dev/lupio/lupio_tty.cc:119]: I personally strongly prefer it. Skipping the braces has caused hard to find bugs in the past.

File: src/dev/lupio/lupio_tty.cc
Comment (powerjg) [reply src/dev/lupio/lupio_tty.cc:65]: My confusion is which console should this post the interrupt to? What if there are two consoles (i.e., in this case with both the uart and the lupioTTY

Comment (joel-porquet) [reply src/dev/lupio/lupio_tty.cc:65]: I guess the current model doesn't allow this scenario. If you wanted to allow it, you'd probably need for `postConsoleInt()` to accept an ID of some sort to indicate which console is raising its IRQ.

That said, the lupio-TTY is replacing the uart now, so we actually don't need the support for two consoles.

**Comment by lhinman19** [comment]
I just rebased so that it now also includes the lupIO-RNG.

File: src/dev/lupio/lupio_tty.cc
Comment (powerjg) [inline src/dev/lupio/lupio_tty.cc:156]: You shouldn't need the `!!` (unless I'm mistaken). The cast from `uint8_t` to `bool` should implicitly work.

```suggestion
            writIntrEn = val & LUPIO_TTY_WBIT;
            readIntrEn = val & LUPIO_TTY_RBIT;
```

**Review by powerjg — COMMENTED** [review]
One more tiny thing, then it's good to go from my end.

File: src/dev/lupio/lupio_tty.cc
Comment (joel-porquet) [reply src/dev/lupio/lupio_tty.cc:156]: It's a pretty common idiom in the Linux kernel, which I think comes from the fact that the kernel doesn't use `bool` very often (but `int` instead). Here you're right, the conversion to bool should work.

**Pull request merged**
